cpp/tensorrt_llm/kernels/cuteDslKernels/moeUtils.cu

```diff
@@ -0,0 +1,436 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "tensorrt_llm/common/assert.h"
+#include "tensorrt_llm/common/cudaUtils.h"
+#include "tensorrt_llm/common/envUtils.h"
+#include "tensorrt_llm/kernels/cuteDslKernels/moeUtils.h"
+#include "tensorrt_llm/kernels/quantization.cuh"
+#include "tensorrt_llm/kernels/quantization.h"
+
+#include <cuda_fp4.h>
+#include <cute/numeric/numeric_types.hpp>
+
+namespace tensorrt_llm::kernels::cute_dsl
+{
+namespace
+{
+using ElemCopyType = uint4;
+using SFCopyType = uint32_t;
+
+template <typename T>
+auto constexpr bitsPerElem()
+{
+#ifdef ENABLE_FP4
+    return std::is_same_v<T, __nv_fp4_e2m1> ? 4 : cute::sizeof_bits_v<T>;
+#else
+    return cute::sizeof_bits_v<T>;
+#endif
+}
+
+template <typename T>
+auto constexpr elemPerCopy()
+{
+    return bitsPerElem<ElemCopyType>() / bitsPerElem<T>();
+}
+
+template <typename T>
+auto constexpr sfElemPerCopy()
+{
+    return bitsPerElem<SFCopyType>() / bitsPerElem<T>();
+}
+} // namespace
+
+template <typename InputType, typename SFType, int32_t kSFVecSize, int32_t kThreadsPerBlock>
+__global__ void moePermuteKernel(InputType const* input, InputType* permuted_output, SFType const* input_sf,
```

Comment:
Is there a particular reason we need to have custom implementation here rather than using the existing cutlass kernels like in `cpp/tensorrt_llm/thop/moeUtilOp.cpp`